File: server/php/src/HostedCheckout/HostedCheckoutController.php
<?php
namespace MyApp\HostedCheckout;

use MyApp\Payment\PaymentDetailsService;
use Slim\Psr7\Request;
use Slim\Psr7\Response;
use Laminas\Diactoros\Response\JsonResponse;
use InvalidArgumentException;

class HostedCheckoutController
{
    private $mapper;
    private $hostedCheckoutService;
    private $paymentDetailsService;

    public function __construct(
        HostedCheckoutMapper $mapper,
        HostedCheckoutService $hostedCheckoutService,
        PaymentDetailsService $paymentDetailsService
    ) {
        $this->mapper = $mapper;
        $this->hostedCheckoutService = $hostedCheckoutService;
        $this->paymentDetailsService = $paymentDetailsService;
    }

    public function getCreateHostedCheckout(Request $request, Response $response, $args)
    {
        $dto = $this->mapper->toEmptyBasicDto();
        return new JsonResponse($dto, 200, [], true);
    }

    public function getPaymentResponse(Request $request, Response $response, $args)
    {
        $hostedCheckoutId = $request->getQueryParams()['hostedCheckoutId'];

        if (empty($hostedCheckoutId)) {
            throw new InvalidArgumentException("Hosted checkout id is missing!");
        }

        $paymentDetailsResponse = $this->paymentDetailsService->getPaymentDetailsForHostedCheckout($hostedCheckoutId);
        return new JsonResponse($paymentDetailsResponse->toObject(), 200, [], true);
    }

    public function createHostedCheckout(Request $request, Response $response, $args) 
    {
        $createHostedCheckoutRequest = $this->mapper->toCreateHostedCheckoutRequest($this->mapper->toCreateHostedCheckoutBasicDto($request));
        $createHostedCheckoutResponse = $this->hostedCheckoutService->createHostedCheckoutResponse($createHostedCheckoutRequest);

        return new JsonResponse($createHostedCheckoutResponse->toObject(), 200, [], true);   
    }
}

File: server/java/src/main/java/com/worldlinesolutions/onlinepayments/example/hostedcheckout/HostedCheckoutController.java
package com.worldlinesolutions.onlinepayments.example.hostedcheckout;

import com.onlinepayments.domain.CreateHostedCheckoutResponse;
import com.onlinepayments.domain.PaymentDetailsResponse;
import com.worldlinesolutions.onlinepayments.example.payment.PaymentDetailsService;
import io.micrometer.common.util.StringUtils;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/hostedcheckout")
public class HostedCheckoutController {

    private final HostedCheckoutMapper hostedCheckoutMapper;
    private final HostedCheckoutService hostedCheckoutService;
    private final PaymentDetailsService paymentDetailsService;

    public HostedCheckoutController(
            HostedCheckoutMapper hostedCheckoutMapper,
            HostedCheckoutService hostedCheckoutService,
            PaymentDetailsService paymentDetailsService
    ) {
        this.hostedCheckoutMapper = hostedCheckoutMapper;
        this.hostedCheckoutService = hostedCheckoutService;
        this.paymentDetailsService = paymentDetailsService;
    }

    @GetMapping
    public CreateHostedCheckoutBasicDto getHostedCheckout() {
        return hostedCheckoutMapper.toEmptyDto();
    }

    @GetMapping("/outcome")
    public PaymentDetailsResponse getPaymentResponse(
            @RequestParam("RETURNMAC") String returnMac,
            @RequestParam("hostedCheckoutId") String hostedCheckoutId
    ) {
        if (StringUtils.isEmpty(hostedCheckoutId))
            throw new IllegalArgumentException("Hosted checkout id is missing!");

        return paymentDetailsService.getPaymentDetailsForHostedCheckout(hostedCheckoutId);
    }

    @PostMapping("/basic")
    public CreateHostedCheckoutResponse createHostedCheckout(@RequestBody CreateHostedCheckoutBasicDto createHostedCheckoutBasicDto) {
        return hostedCheckoutService.createHostedCheckoutResponse(
                hostedCheckoutMapper.toCreateHostedCheckoutRequest(createHostedCheckoutBasicDto)
        );
    }

}

